SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialLatencyTracer.cpp

```diff
@@ -0,0 +1,360 @@
+	{
+		UE_LOG(LogSpatialLatencyTracing, Warning, TEXT("(%s) : Failed to create Actor/Func trace (%s)"), *WorkerId, *TraceDesc);
+		return false;
+	}
+
+	FString SpanMsg = FormatMessage(TraceDesc);
+	TraceSpan NewTrace = improbable::trace::Span::StartSpan(TCHAR_TO_UTF8(*SpanMsg), nullptr);
+
+	WriteKeyFrameToTrace(&NewTrace, FString::Printf(TEXT("Begin trace : %s"), *FunctionName));
+
+	TraceMap.Add(Key, MoveTemp(NewTrace));
+
+	return true;
+}
+
+bool USpatialLatencyTracer::ContinueLatencyTrace_Internal(const AActor* Actor, const FString& FunctionName, const FString& TraceDesc)
+{
+	FScopeLock Lock(&Mutex);
+
+	TraceSpan* ActiveTrace = GetActiveTrace();
+	if (ActiveTrace == nullptr)
+	{
+		UE_LOG(LogSpatialLatencyTracing, Warning, TEXT("(%s) : No active trace to continue (%s)"), *WorkerId, *TraceDesc);
+		return false;
+	}
+
+	TraceKey Key = CreateNewTraceEntry(Actor, FunctionName);
+	if (Key == InvalidTraceKey)
+	{
+		UE_LOG(LogSpatialLatencyTracing, Warning, TEXT("(%s) : Failed to create Actor/Func trace (%s)"), *WorkerId, *TraceDesc);
+		return false;
+	}
+
+	WriteKeyFrameToTrace(ActiveTrace, TCHAR_TO_UTF8(*TraceDesc));
+	WriteKeyFrameToTrace(ActiveTrace, FString::Printf(TEXT("Continue trace : %s"), *FunctionName));
+
+	TraceMap.Add(Key, MoveTemp(*ActiveTrace));
+	TraceMap.Remove(ActiveTraceKey);
+
+	return true;
+}
+
+bool USpatialLatencyTracer::EndLatencyTrace_Internal()
+{
+	FScopeLock Lock(&Mutex);
+
+	TraceSpan* ActiveTrace = GetActiveTrace();
+	if (ActiveTrace == nullptr)
+	{
+		UE_LOG(LogSpatialLatencyTracing, Warning, TEXT("(%s) : No active trace to end"), *WorkerId);
+		return false;
+	}
+
+	WriteKeyFrameToTrace(ActiveTrace, TEXT("End Trace"));
+	ActiveTrace->End();
+
+	TraceMap.Remove(ActiveTraceKey);
+
+	return true;
+}
+
+TraceKey USpatialLatencyTracer::CreateNewTraceEntry(const AActor* Actor, const FString& FunctionName)
+{
+	static TraceKey NextTraceKey = 1;
+
+	if (UClass* ActorClass = Actor->GetClass())
+	{
+		if (UFunction* Function = ActorClass->FindFunctionByName(*FunctionName))
+		{
+			ActorFuncKey Key{ Actor, Function };
+			if (TrackingTraces.Find(Key) == nullptr)
+			{
+				TrackingTraces.Add(Key, NextTraceKey);
+				return NextTraceKey++;
+			}
+			UE_LOG(LogSpatialLatencyTracing, Warning, TEXT("(%s) : ActorFunc already exists for trace"), *WorkerId);
+		}
+	}
+
+	return InvalidTraceKey;
+}
+
+USpatialLatencyTracer::TraceSpan* USpatialLatencyTracer::GetActiveTrace()
+{
+	return TraceMap.Find(ActiveTraceKey);
+}
+
+void USpatialLatencyTracer::WriteKeyFrameToTrace(const TraceSpan* Trace, const FString& TraceDesc)
+{
+	if (Trace != nullptr)
+	{
+		FString TraceMsg = FormatMessage(TraceDesc);
+		improbable::trace::Span::StartSpan(TCHAR_TO_UTF8(*TraceMsg), Trace).End();
+	}
+}
+
+FString USpatialLatencyTracer::FormatMessage(const FString& Message) const
+{
+	return FString::Printf(TEXT("(%s) : %s"), *WorkerId.Left(18), *Message);
+}
+
+#endif // TRACE_LIB_ACTIVE
+
+void USpatialLatencyTracer::Debug_SendTestTrace()
```

Comment:
Could we add comments on what this is and what is meant to be observed. Is it something that could be a unit test?

Reply:
There's a comment in the header file. Is that not enough?

Reply:
I don't really know what it means other than 3 things happen, which I can see from code, but I still don't know what they do. This might be part of a broader comment that we don't really define what a trace is in comments anywhere and it's not immediately obvious from the API.
But it still looks like we have a unit test embedded in the class.

Reply:
As this relies on invoking a third-party library, this isn't easily unit-testable. I'll add a comment about what a trace is.

Reply:
We could have a test module that depends on it. Although if we aren't putting this into a module then it would be a fair amount of effort.